metadata-ingestion/src/datahub/utilities/file_backed_collections.py

```diff
@@ -1,87 +1,162 @@
         # a poor-man's LRU cache.
         self._active_object_cache = collections.OrderedDict()
 
-        # These settings are optimized for performance.
-        # See https://www.sqlite.org/pragma.html for more information.
-        # Because we're only using this file to offload data from memory, we don't need
-        # to worry about data integrity too much.
-        self._conn.execute('PRAGMA locking_mode = "EXCLUSIVE"')
-        self._conn.execute('PRAGMA synchronous = "OFF"')
-        self._conn.execute('PRAGMA journal_mode = "MEMORY"')
-        self._conn.execute(f"PRAGMA journal_size_limit = {100 * 1024 * 1024}")  # 100MB
-
-        # The key will automatically be indexed.
+        # Create the table. We're not using "IF NOT EXISTS" because creating
+        # the same table twice indicates a client usage error.
         self._conn.execute(
-            "CREATE TABLE IF NOT EXISTS data (key TEXT PRIMARY KEY, value BLOB)"
+            f"""CREATE TABLE {self.tablename} (
+                key TEXT PRIMARY KEY,
+                value BLOB
+                {''.join(f', {column_name} BLOB' for column_name in self.extra_columns.keys())}
+            )"""
         )
 
+        # The key column will automatically be indexed, but we need indexes
+        # for the extra columns.
+        for column_name in self.extra_columns.keys():
+            self._conn.execute(
+                f"CREATE INDEX {self.tablename}_{column_name} ON {self.tablename} ({column_name})"
+            )
+
     def _add_to_cache(self, key: str, value: _VT) -> None:
         self._active_object_cache[key] = value
 
-        if (
-            len(self._active_object_cache)
-            > self.cache_max_size + self.cache_eviction_batch_size
-        ):
-            num_items_to_prune = len(self._active_object_cache) - self.cache_max_size
+        if len(self._active_object_cache) > self.cache_max_size:
+            # Try to prune in batches rather than one at a time.
+            num_items_to_prune = min(
+                len(self._active_object_cache), self.cache_eviction_batch_size
+            )
             self._prune_cache(num_items_to_prune)
 
     def _prune_cache(self, num_items_to_prune: int) -> None:
-        items_to_write: List[Tuple[str, SqliteValue]] = []
+        items_to_write: List[Tuple[SqliteValue, ...]] = []
         for _ in range(num_items_to_prune):
             key, value = self._active_object_cache.popitem(last=False)
-            items_to_write.append((key, self.serializer(value)))
+
+            values = [key, self.serializer(value)]
+            for column_serializer in self.extra_columns.values():
+                values.append(column_serializer(value))
+            items_to_write.append(tuple(values))
 
         self._conn.executemany(
-            "INSERT OR REPLACE INTO data (key, value) VALUES (?, ?)", items_to_write
+            f"""INSERT OR REPLACE INTO {self.tablename} (
```

Comment:
So we only drop rows onto disk once the cache max size is exceeded?

Reply:
In the future, I can see us wanting to bypass this behavior and always dual write to disk (maybe that's what we are already doing though)

Reply:
(Or is the caller expected to just flush?)